### PR TITLE
Initial commit of mkfringeflat.py, response.py and findloc.py

### PR DESCRIPTION
**Pull request by stscirij**
Phil wrote findloc.py

### CODE REVIEW

File: stistools/mkfringeflat.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+            rmsvalues[i] = sigma/mean
+            print("shift = {}, rms = {}".format(current_shift[i], rmsvalues[i]))
+
+        #
+        # Determine shift that delivers the best RMS by an inverse-weighted average
+        weight = 1.0/(rmsvalues*rmsvalues)
+        rownum = range(nshifts)
+        weighted_rownum = weight*rownum
+        minrms = rmsvalues.min()
+        min_row = np.where(rmsvalues == minrms)[0][0]
```
Comment (philhodge) [inline stistools/mkfringeflat.py]: Would `np.argmin()` be useful here?

File: stistools/mkfringeflat.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+        if min_row == 0 or min_row == nshifts - 1:
+            print(" ")
+            print("WARNING: Best shift found on the edge of the specified shift range.")
+            print("You are advised to try again after adjusting the shift range accordingly")
+            theshift = current_shift[min_row]
+
+        else:
+            if min_row >= 2 and min_row <= nshifts - 3:
+                first_row = min_row - 2
```
Comment (philhodge) [inline stistools/mkfringeflat.py]: Should the "- 2" be "- 3" (one less than iraf)?

File: stistools/mkfringeflat.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+                last_row = min_row + 3
```
Comment (philhodge) [inline stistools/mkfringeflat.py]: Should the "+ 3" be "+ 2"?

File: stistools/mkfringeflat.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+            elif min_row == 1 or min_row == nshifts - 2:
+                first_row = min_row - 1
```
Comment (philhodge) [inline stistools/mkfringeflat.py]: Should "- 1" be "- 2" (same as iraf)?

File: stistools/mkfringeflat.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+                last_row = min_row + 2
```
Comment (philhodge) [inline stistools/mkfringeflat.py]: Should "+ 2" be "+ 1"?

File: stistools/mkfringeflat.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+        last_row = min_row + 3
+        if min_row == 0 or min_row == nscales - 1:
+            print(" ")
+            print("WARNING: Best scale found on the edge of the specified scale range.")
+            print("You are advised to try again after adjusting the scale range accordingly")
+            thescale = current_scale[min_row]
+
+        else:
+            if min_row >= 2 and min_row <= nscales - 3:
+                first_row = min_row - 2
```
Comment (philhodge) [inline stistools/mkfringeflat.py]: Should "- 2" be "- 3" (one less than iraf)?

File: stistools/mkfringeflat.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+                last_row = min_row + 3
```
Comment (philhodge) [inline stistools/mkfringeflat.py]: Should "+ 3" be "+ 2" (same as iraf)?

File: stistools/mkfringeflat.py [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+            elif min_row == 1 or min_row == nscales - 2:
+                first_row = min_row - 1
```
Comment (philhodge) [inline stistools/mkfringeflat.py]: Should "- 1" be "- 2"?

File: stistools/mkfringeflat.py [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+                last_row = min_row + 2
```
Comment (philhodge) [inline stistools/mkfringeflat.py]: Should "+ 2" be "+ 1"?

**Comment by sean-lockwood** [comment]
You might copy/modify the doc string from here:
https://github.com/spacetelescope/stistools/blob/dev_defringe/stistools/defringe/mkfringeflat.py

File: stistools/mkfringeflat.py [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+
+            star_cont_shift = scidata / shifted_flat
+
+            if "G750M" in opt_elem:
+                norm_star_cont_shift = response(star_cont_shift[fline:lline,colstart:colstop],
+                                                star_cont_shift[fline:lline,colstart:colstop],
+                                                threshold=1.0e-20,
+                                                function="spline3",
+                                                sample="*",
+                                                naverage=2,
```
Comment (sean-lockwood) [inline stistools/mkfringeflat.py]: `naverage` doesn't get used below, which implies it should be 1 and anything else should raise a `NotImplementedError`.

File: stistools/response.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+import numpy as np
+from astropy.io import fits
+
+import fit1d
+
+def response(calibration_array, normalization_array,
+             threshold=None,
+             function="spline3",
+             sample='*',
+             naverage=2,
```
Comment (sean-lockwood) [inline stistools/response.py]: `naverage` doesn't get used below, which implies it should be 1 and anything else should raise a `NotImplementedError`.

File: stistools/response.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    """
+
+    return 1
+
+def normalise_response(calibration_array, dispaxis, threshold, fitted_spectrum):
+    """Like re_normalize in response.x
+
+    """
+
+    nrows, ncols = calibration_array.shape
```
Comment (sean-lockwood) [inline stistools/response.py]: `nrows` and `ncols` not used.

File: stistools/response.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+def get_dispaxis(calibration_array):
+    """For completeness, include a function to get this.  For our purposes, dispaxis is
+    always 1
+
+    """
+
+    return 1
+
+def normalise_response(calibration_array, dispaxis, threshold, fitted_spectrum):
```
Comment (sean-lockwood) [inline stistools/response.py]: `dispaxis` not actually implemented.  Use a `NotImplementedError` to catch a non-default value?

File: stistools/response.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
+import math
```
Comment (sean-lockwood) [inline stistools/response.py]: `math` and `fits` aren't used in this file.

File: stistools/response.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+def response(calibration_array, normalization_array,
+             threshold=None,
+             function="spline3",
+             sample='*',
+             naverage=2,
+             order=10,
+             low_reject=3.0,
+             high_reject=3.0,
+             niterate=2,
+             grow=0.0):
```
Comment (sean-lockwood) [inline stistools/response.py]: Similarly, `grow != 0.` should raise a `NotImplementedError` in `fit1d()`.

File: stistools/response.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
+import math
+import numpy as np
+from astropy.io import fits
+
+import fit1d
```
Comment (sean-lockwood) [inline stistools/response.py]: This import won't work with the installed package.

Maybe use something like:
`from .fit1d import fit1d`

**Comment by sean-lockwood** [comment]
I've done some initial simulations of the `response` function here:
`/grp/hst/stis/defringe/response/test_response.*`

For the normalization to function properly, there's a window of valid extraction heights that appears to be tied to the S/N of the source, and a possible dependence on the trace throwing data out of the window.  Perhaps we can do more detailed simulations to constrain the calls to `response`?

File: stistools/findloc.py [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    shape = input.shape
+    # first column in a slice
+    fcol = int(round(low_frac * shape[1]))
+    # last column in a slice
+    lcol = int(round(high_frac * shape[1])) + 1
+
+    # first line in a slice
+    fline = int(round(low_line_frac * shape[0]**2 / 1024.)) + 1
+    # last line in a slice
+    lline = shape[1] - fline
```
[the remark this answers is not in the document]
Comment (philhodge) [reply stistools/findloc.py]: No, the expression for `lline` should be `lline = shape[0] - fline - 2`.

**Comment by stscirij** [comment]
Still haven't addressed most of the review comments, merging so others can try it out

**Pull request merged**